src/python/pants/option/subsystem.py

```diff
@@ -119,9 +124,80 @@ def __eq__(self, other: Any) -> bool:
         return bool(self.options == other.options)
 
 
-_T = TypeVar("_T", bound=Subsystem)
+_SubsystemT = TypeVar("_SubsystemT", bound=Subsystem)
+_T = TypeVar("_T")
 
 
-async def _construct_subsytem(subsystem_typ: type[_T]) -> _T:
+class Option(Generic[_T]):
+    """Descriptor for subsystem options.
+
+    This class serves two purposes:
+        - Register your subsystem's options
+        - Provide a typed property for Python usage
+
+    Usage:
+        class Engine(Subsystem):
+            ...
+
+            cylinders = Option[int]("--cylinders", default=6, help="...")
+
+        ...
+
+        engine: Engine = ...
+        engine.cylinders  # mypy knows this is an int
+
+    Under-the-hood:
+        - You can pass a `converter` function to convert the option value into the property value
+            E.g. `converter=tuple`
+    """
+
+    # NB: We have to ignore type because we can't `cast(_T, x)` as `_T` is purely a type-checking
+    # construct and `cast()` is a runtime function.
+    DEFAULT_CONVERTER: Callable[[Any], _T] = lambda x: x  # type: ignore
+
+    def __init__(
+        self,
+        *args: str,
+        converter: Callable[[Any], _T] = DEFAULT_CONVERTER,
+        **kwargs: Any,
+    ):
+        self.args = args
+        self.kwargs = kwargs
+        self._converter = converter
+
+    @overload
+    def __get__(self, obj: None, objtype: type[_SubsystemT]) -> Option:
+        ...
+
+    @overload
+    def __get__(self, obj: _SubsystemT, objtype: type[_SubsystemT]) -> _T:
+        ...
+
+    def __get__(self, obj: _SubsystemT | None, objtype: type[_SubsystemT]) -> Option | _T:
+        assert issubclass(
+            objtype, Subsystem
+        ), "Option should only be used as attributes of a Subsystem"
+        if obj is None:
+            return self
+        long_name = self.args[-1]
+        option_value = getattr(obj.options, long_name[2:].replace("-", "_"))
+        return self._converter(option_value)
+
+
+BoolOption: functools.partial[Option[bool]] = functools.partial(Option, type=bool)
+StrOption: functools.partial[Option[str]] = functools.partial(Option, type=str)
+IntOption: functools.partial[Option[int]] = functools.partial(Option, type=int)
+# NB: You'll still need to prive help=""
+ArgsOption: functools.partial[Option["tuple[str, ...]"]] = functools.partial(
+    Option,
+    "--args",
+    type=list,
+    member_type=shell_str,
+    passthrough=True,
+    converter=tuple,
+)
```

Comment:
I have some ideas for how to make this even better with MyPy by having `BoolOption` etc subclass `Option` and changing `__init__()` etc to say exactly what args are necessary and with what types.

Unlike the Target API, you would not subclass these "templates" to define a new option - you would still use your very ergonomic approach. Only the templates would be subclasses.

But this is totally fine for now, especially because `__init__` is untyped still.

Reply:
A good example of what benefits I mean: `StringOption` can have in its constructor `default: str | None = None`.